phoenix-core/src/main/java/org/apache/phoenix/compile/ProjectionCompiler.java

```diff
@@ -497,42 +502,98 @@ public static RowProjector compile(StatementContext context, SelectStatement sta
+                        }
+                    }};
+            Map<String, List<KeyValueColumnExpression>>
+                    serverAttributeToKVExpressionMap =
+                    new HashMap<String, List<KeyValueColumnExpression>>() {{
+                        for (String attribute : attributeToFunctionMap.keySet()) {
+                            put(attribute, new ArrayList<>());
+                        }
+                    }};
+            int counter = 0;
+            for (String attribute : scanAttributes) {
+                for (int i = 0; i < serverParsedKVFuncs.size(); i++) {
+                    if (attributeToFunctionMap.get(attribute)
+                            .isInstance(serverParsedKVFuncs.get(i))) {
+                        initialToShuffledPositionMap.put(i, counter++);
+                        serverAttributeToFuncExpressionMap.get(attribute)
+                                .add(serverParsedKVFuncs.get(i));
+                        serverAttributeToKVExpressionMap.get(attribute)
+                                .add(serverParsedKVRefs.get(i));
+                    }
+                }
+            }
+            for (Map.Entry<String, Class> entry : attributeToFunctionMap.entrySet()) {
+                if (serverAttributeToFuncExpressionMap.get(entry.getKey()).size() > 0) {
+                    serializeServerParsedExpressionInformationAndSetInScan(context, entry.getKey(),
+                            serverAttributeToFuncExpressionMap.get(entry.getKey()),
+                            serverAttributeToKVExpressionMap.get(entry.getKey()));
+                }
+            }
             KeyValueSchemaBuilder builder = new KeyValueSchemaBuilder(0);
-            for (Expression expression : arrayKVRefs) {
+            for (Expression expression : serverParsedKVRefs) {
                 builder.addField(expression);
             }
             KeyValueSchema kvSchema = builder.build();
             ValueBitSet arrayIndexesBitSet = ValueBitSet.newInstance(kvSchema);
             builder = new KeyValueSchemaBuilder(0);
-            for (Expression expression : arrayKVFuncs) {
+            for (Expression expression : serverParsedKVFuncs) {
                 builder.addField(expression);
             }
             KeyValueSchema arrayIndexesSchema = builder.build();
 
             Map<Expression, Expression> replacementMap = new HashMap<>();
-            for(int i = 0; i < arrayOldFuncs.size(); i++){
-                Expression function =arrayKVFuncs.get(i);
-                replacementMap.put(arrayOldFuncs.get(i), new ArrayIndexExpression(i, function.getDataType(), arrayIndexesBitSet, arrayIndexesSchema));
+            for (int i = 0; i < serverParsedOldFuncs.size(); i++) {
+                Expression function = serverParsedKVFuncs.get(i);
+                replacementMap.put(serverParsedOldFuncs.get(i),
+                        new ArrayIndexExpression(initialToShuffledPositionMap.get(i),
```

Comment:
I didn't quite understand why we are creating ArrayIndexExpression here ? Does this code not apply to json expressions ?

Reply:
This applies to JSON Expressions too. At a logical level ARRAY_ELM(array, Index) and JSON_FUN(root, path) are identical in structure and evaluation path. We are reusing it. Actually I had a local change to introduce JSONPathExpression and create different code path and that needed multiple conditional checks at multiple places. So continued with this. What's your recommendation?

Reply:
I created a JSONPathExpression subclass of ArrayIndexExpression to make it clear. Hope this will make things clear